User status refactor

wyattjoh opened the pull request:
# User status refactor 

## What does this PR do?

- Adds user status refactors to improve the user management flow
- Adds user action counts to optimize admin actions
- Added support for deflate + brotli compression of static assets
- Moved static asset root from `/` and `/client` to `/static`
  - `/embed.js` is now `/static/embed.js`
  - `/client/embed/stream/bundle.js` is now `/static/embed/stream/bundle.js`
  - etc.

### Changes on the Client
- Refactor of Community. Deprecation of REST endpoints. Now entirely working with the Graph
- Moving User Actions: Banning and Suspending to the User Detail Drawer
- Refactoring changes to match the new user status and functionality
- Adds User History


## Preparing this PR

1. Run pending migrations with `./bin/cli -c .env migration run`
2. Update and verify counts in the database `./bin/cli -c .env verify db --fix`
3. `yarn dev-start` and `yarn build-watch` has been replaced with `yarn watch`. This combines the `yarn dev-start` and `yarn build-watch` into a single command!

## How do I test this PR?

- Verify user can be suspended, banned, and flagged/rejected/changed/approved

## Todo

- [x] Send suspend email
- [x] Send ban email
- [x] Send username rejection email
- [ ] Add graph filtering for users

wyattjoh committed:
perms cleanup

wyattjoh committed:
Merge branch 'master' into user-status-refactor

wyattjoh committed:
initial pass at status support

wyattjoh committed:
added prebuild

wyattjoh committed:
added username edit, tests

wyattjoh committed:
fixed changed package script

wyattjoh committed:
added more mutations

wyattjoh committed:
user queries

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
added migration, migration tests :tada:

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
added emails for status changes

alinhle commented on bin/cli-users:
```diff
@@ -106,20 +106,17 @@ async function createUser(options) {
     }
 
-    util.shutdown();
+    await UsersService.sendEmailConfirmation(user, answers.email.trim());
```
👍

wyattjoh committed:
added action counts for users

wyattjoh committed:
staff members cannot be flagged

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
spelling

okbel committed:
Missing proptypes

okbel committed:
Merge branch 'master' of github.com:coralproject/talk into mod-improv

okbel committed:
Account History

okbel committed:
styling

okbel committed:
Merge branch 'user-status-refactor' of github.com:coralproject/talk into mod-improv

okbel committed:
Client Refactor

okbel committed:
Refactor Admin Side queries

okbel committed:
wip

okbel commented:
I dropped the DB and ran the install and it threw this error. 

```
Migration 1510174676_user_status.js failed
{ MongoError: Invalid Operation, no operations specified
    at Function.MongoError.create (/Users/bcurcio/Develop/talk/node_modules/mongodb-core/lib/error.js:31:11)
    at toError (/Users/bcurcio/Develop/talk/node_modules/mongodb/lib/utils.js:139:22)
    at UnorderedBulkOperation.execute (/Users/bcurcio/Develop/talk/node_modules/mongodb/lib/bulk/unordered.js:512:27)
    at Object.up (/Users/bcurcio/Develop/talk/migrations/1510174676_user_status.js:208:16)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  name: 'MongoError',
  message: 'Invalid Operation, no operations specified',
  driver: true }
POST /api/v1/setup 500 79.650 ms - 2
```

I get the same error when if I try to run the migration after it. And because the `/setup` responded 500 I don't have a valid instance now and I can't move forward with the install.

okbel commented:
I used an old DB and ran the migrations, and it's working. (Not the case when I use a clean db 🤔 )

Now I'm getting this error:
`Invalid options provided to ApolloServer:  are not valid permissions.`

wyattjoh committed:
fixed migration issues

wyattjoh committed:
Merge branch 'next' into user-status-refactor

okbel commented on graph/mutators/user.js:
```diff
@@ -1,29 +1,34 @@
-  const result = await UsersService.rejectUsername(id, message);
-  if (result) {
-    pubsub.publish('usernameRejected', result);
+const setUserSuspensionStatus = async (ctx, id, until) => {
```
This fn used to take the message. Where should the message go?

okbel replied:
`const suspendUser = async ({pubsub}, {id, message, until}) => {...}`

okbel committed:
Suspend User Mutation

okbel committed:
Removing setCommenterStatus

okbel committed:
Refactor for setUserStatus to setUserBanStatus

okbel committed:
Refactor for setUserStatus to setUserBanStatus

okbel committed:
refetch status

okbel committed:
Banned User Stream

okbel committed:
Missing proptype

okbel committed:
bug error view options

okbel committed:
Banned User Stream

okbel committed:
Safe access

okbel committed:
removing pending

okbel committed:
updated e2e

okbel committed:
Safe access to objects

okbel committed:
Reject username

okbel commented:
Found by the tests: 

Performing this: 
```
mutation RejectUsername {
  rejectUsername(id: "3ceb1835-2db5-4268-9c83-ff2318021ec0") {
    errors {
      __typename
    }
  }
}

query User {
  user(id: "3ceb1835-2db5-4268-9c83-ff2318021ec0") {
    state {
      status {
        username {
          status
        }
      }
    }
  }
}
```
`RejectUsername` does not modify the state/status of the user. State keeps comming `null`.

@wyattjoh 
#blocker

okbel commented:
I've found another bug with suspended users:  

Auth endpoint is returning `state` full, with a suspension while `me` in the graph returns `null` this is as soon as I log in with a suspended user

<img width="700" alt="screen shot 2017-11-20 at 12 16 05 pm" src="https://user-images.githubusercontent.com/1401559/33025593-2fa5fb58-cded-11e7-9288-8a46c63eef29.png">

<img width="700" alt="screen shot 2017-11-20 at 12 16 13 pm" src="https://user-images.githubusercontent.com/1401559/33025585-2a01e626-cded-11e7-9ccd-a9e7bace9f69.png">

wyattjoh committed:
fixed query issues

okbel committed:
setUsername refactor

okbel committed:
setUsername refactor

okbel committed:
Now using changeUsername

wyattjoh committed:
fixes to cli

wyattjoh committed:
added USERS_NEW event to SSE bus

okbel committed:
changes ChangeUsername

okbel committed:
Bug suspending/banning/rejecting users,

okbel committed:
user should not be suspended if username is rejecte

okbel committed:
wip ChangeUsername

okbel committed:
wip

okbel committed:
updated tests, changeUsername container, etc

okbel committed:
refactor changeUsername - Product question,

okbel committed:
removing message

okbel committed:
Updating contations and deleting legacy code

okbel committed:
Hiding messages for suspension and bann

okbel committed:
Updating componets

okbel committed:
Updating tests: dont wait for message

wyattjoh committed:
added message to ban/suspend workflows

okbel committed:
banUser and unBanUser wip - removing setStatus

okbel committed:
banUser and unBanUser - removing setUserStatus

okbel committed:
Merge branch 'mod-improv' of github.com:coralproject/talk into user-status-refactor

okbel committed:
Adding message

okbel committed:
Restoring message for suspend

okbel commented:
Using this query to retrieve user history data:

```gql
{
  user(id: "e75b9bb8-8e6c-4ffc-ae5a-ed0ed9c257ae") {
    state {
      status {
        suspension {
          history {
            until
            created_at
            assigned_by {
              username
            }
          }
        }
        banned {
          history {
            status
            assigned_by {
              username
            }
            created_at
          }
        }
        username {
          history {
            status
            assigned_by {
              username
            }
            created_at
          }
        }
      }
    }
  }
}
```

Returns this error: `Cannot return null for non-nullable field User.username.` multiple times.

http://localhost:3000/api/v1/graph/iql?query=%7B%0A%20%20user(id%3A%20%22e75b9bb8-8e6c-4ffc-ae5a-ed0ed9c257ae%22)%20%7B%0A%20%20%20%20state%20%7B%0A%20%20%20%20%20%20status%20%7B%0A%20%20%20%20%20%20%20%20suspension%20%7B%0A%20%20%20%20%20%20%20%20%20%20history%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20until%0A%20%20%20%20%20%20%20%20%20%20%20%20created_at%0A%20%20%20%20%20%20%20%20%20%20%20%20assigned_by%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20username%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20banned%20%7B%0A%20%20%20%20%20%20%20%20%20%20history%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20status%0A%20%20%20%20%20%20%20%20%20%20%20%20assigned_by%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20username%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20created_at%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20username%20%7B%0A%20%20%20%20%20%20%20%20%20%20history%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20status%0A%20%20%20%20%20%20%20%20%20%20%20%20assigned_by%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20username%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20created_at%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D&operationName=null
<img width="700" alt="screen shot 2017-11-23 at 8 28 37 am" src="https://user-images.githubusercontent.com/1401559/33170797-5ff400ba-d028-11e7-8cd0-e9f7e0932629.png">

okbel committed:
Suspension response

wyattjoh committed:
added support for user assigned_by fields to get resolved

wyattjoh committed:
added role management to graph

wyattjoh committed:
removed double await

wyattjoh committed:
fixes based on feedback

okbel committed:
missing fragment

okbel committed:
Adding diff and fix to utils

okbel committed:
Working account history

okbel committed:
Build User History

okbel committed:
UserDetail Refactor

okbel committed:
UserDetail Refactor

okbel committed:
Adding addUserRole and removeUserRole

kgardnr commented:
@okbel @wyattjoh I logged these bugs related to errors I'm seeing on this branch - 

- https://www.pivotaltracker.com/story/show/153174776
- https://www.pivotaltracker.com/story/show/153174892

okbel committed:
Adding UserDetailCommentList

okbel committed:
Approve Username fix

wyattjoh committed:
roles -> role, null -> COMMENTER for user roles

wyattjoh committed:
fixed bug with featuring comments doesn't work

wyattjoh committed:
more role fixes

okbel committed:
Adding ban message and setRole

okbel committed:
Ban user dialog steps

wyattjoh committed:
Merge branch 'next' into user-status-refactor

okbel committed:
Translations and Styling

okbel committed:
Merge branch 'user-status-refactor' of github.com:coralproject/talk into user-status-refactor

okbel committed:
Wrong filename

okbel committed:
Removing ternary

okbel committed:
Updating after approving a user

okbel committed:
Linting

okbel committed:
Merge branch 'master' of github.com:coralproject/talk into user-status-refactor

okbel committed:
New path

okbel committed:
Updated comment user status path

wyattjoh committed:
added better compression support

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
improved query performance with indexes, added engine for debugging

okbel committed:
Moving ActionsMenu

okbel committed:
Styling and functionality

okbel committed:
Functionality and styling

okbel committed:
UserInfoTooltip, styling and func

okbel committed:
Adding old user status

okbel committed:
Adding old user status

okbel committed:
wip

okbel committed:
Adding react-mdl

okbel commented:
Test are failling giving the following output: @wyattjoh 

```js
  talk:middleware:authentication user was on request +4ms
Error: Schema must be an instance of GraphQLSchema. Also ensure that there are not multiple versions of GraphQL installed in your node_modules directory.
    at invariant (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/jsutils/invariant.js:19:11)
    at Object.validate (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/validation/validate.js:60:72)
    at doRunQuery (/home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:88:38)
    at /home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:22:54
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
Error: Schema must be an instance of GraphQLSchema. Also ensure that there are not multiple versions of GraphQL installed in your node_modules directory.
    at invariant (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/jsutils/invariant.js:19:11)
    at Object.validate (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/validation/validate.js:60:72)
    at doRunQuery (/home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:88:38)
    at /home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:22:54
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  talk:middleware:authentication user was on request +6ms
Error: Schema must be an instance of GraphQLSchema. Also ensure that there are not multiple versions of GraphQL installed in your node_modules directory.
    at invariant (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/jsutils/invariant.js:19:11)
    at Object.validate (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/validation/validate.js:60:72)
    at doRunQuery (/home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:88:38)
    at /home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:22:54
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  talk:db:query db.users.findOne({ id: '94794452-d68d-4d0c-9b3b-b49faa2817b3' }, { fields: {} }) +3s
  talk:middleware:authentication user was on request +3s
Error: Schema must be an instance of GraphQLSchema. Also ensure that there are not multiple versions of GraphQL installed in your node_modules directory.
    at invariant (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/jsutils/invariant.js:19:11)
    at Object.validate (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/validation/validate.js:60:72)
    at doRunQuery (/home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:88:38)
    at /home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:22:54
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 ✖ Expected element <.talk-admin-moderation-container> to be visible - element was not found  - expected "visible" but got: "not found"
    at Section.login (/home/ubuntu/talk/test/e2e/page_objects/admin.js:158:28)
    at Page.navigateAndLogin (/home/ubuntu/talk/test/e2e/page_objects/admin.js:46:33)
    at Object.Admin logs in (/home/ubuntu/talk/test/e2e/specs/02_admin.js:23:15)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

okbel committed:
Updated tests and Drawer component

wyattjoh committed:
Adjusted mongo indexes to support index scanning

wyattjoh committed:
Merge branch 'next' into user-status-refactor

kgardnr committed:
Update CSS classname

wyattjoh committed:
added support for deep route caching, settings cache, template optim

wyattjoh committed:
removed reference

wyattjoh committed:
fixed small bug

wyattjoh committed:
Merge branch 'next' into user-status-refactor

okbel committed:
Merge branch 'actions-user-drawer' of github.com:coralproject/talk into user-status-refactor

okbel committed:
merge

wyattjoh committed:
fixed template errors

wyattjoh committed:
linting patch

okbel committed:
Updated with User-Status-Refactor API

okbel committed:
Adding unSuspendUser mutation

okbel committed:
Updated with User-Status-Refactor API

okbel committed:
Updated with User-Status-Refactor API - User Detail Updated

okbel committed:
unBan Update

okbel committed:
Adding utils, and replacing REST old endpoints

okbel committed:
Styling, refactor and instant update after banning and unbanning from community

okbel committed:
Avoiding refetch from Community

wyattjoh committed:
Fixed import error

wyattjoh committed:
Merge branch 'next' into user-status-refactor

okbel committed:
merge conficts

okbel committed:
Adding role, instead of roles. And disable banning and suspending for the user thats logged in

okbel committed:
Classes for AccountHistory

okbel committed:
Adding Classes to the PageObject

okbel committed:
Test

okbel committed:
Adding Classes

okbel committed:
Updating the store mutations

okbel committed:
Adding Classes

okbel committed:
Updating the store mutations

okbel committed:
Adding tests

okbel committed:
Updated tests

wyattjoh committed:
removed dev requirement for engine

wyattjoh committed:
linting

okbel committed:
rejected usernames should not be able to comment

wyattjoh committed:
added mount path to the endpoint to respect the mounting config

okbel committed:
Updating the store mutations

okbel committed:
bounding fns

okbel committed:
Adding tests

okbel committed:
Merge branch 'user-status-refactor' of github.com:coralproject/talk into user-status-refactor

* 'user-status-refactor' of github.com:coralproject/talk:
  added mount path to the endpoint to respect the mounting config
  linting
  removed dev requirement for engine

okbel committed:
updated tests classes

okbel commented:
Test failling with this output: 
```
Error: Schema must be an instance of GraphQLSchema. Also ensure that there are not multiple versions of GraphQL installed in your node_modules directory.
    at invariant (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/jsutils/invariant.js:19:11)
    at Object.validate (/home/ubuntu/talk/node_modules/apollo-server-core/node_modules/graphql/validation/validate.js:60:72)
    at doRunQuery (/home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:88:38)
    at /home/ubuntu/talk/node_modules/apollo-server-core/dist/runQuery.js:22:54
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 ✖ Expected element <.talk-admin-moderation-container> to be visible - element was not found  - expected "visible" but got: "not found"
    at Section.login (/home/ubuntu/talk/test/e2e/page_objects/admin.js:181:28)
    at Page.navigateAndLogin (/home/ubuntu/talk/test/e2e/page_objects/admin.js:46:33)
    at Object.Admin logs in (/home/ubuntu/talk/test/e2e/specs/02_admin.js:23:15)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)

```

kgardnr committed:
Merge branch 'master' of git+ssh://github.com/coralproject/talk into user-status-refactor

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
fixed graphql error

wyattjoh committed:
hid not logged in error

wyattjoh committed:
fixed dist loading error

wyattjoh committed:
updates to cli

- removed -c,--config flag in favor of env only or .env files
- loads .env always if found
- scripts re-written to call cli-serve directly

wyattjoh committed:
fixed tests

wyattjoh committed:
fixed tests
